### PR TITLE
Expose topic filter to command line (addresses #342)

### PR DESCRIPTION
**Pull request by mabelzhang**
Addresses #342. Exposes reader topic filter to command line playback.

The test added in `test_rosbag2_play_end_to_end.cpp` depends on #362.
Usage:
```
ros2 bag play cdr_test --topics /test_topic
ros2 bag play --topics /array_topic -- cdr_test
```

### CODE REVIEW

File: ros2bag/ros2bag/verb/play.py [hunk 1]
```diff
@@ -24,7 +24,8 @@ class PlayVerb(VerbExtension):
 
     def add_arguments(self, parser, cli_name):  # noqa: D102
         parser.add_argument(
-            'bag_file', help='bag file to replay')
+            'input', type=str,
```
Comment (piraka9011) [inline ros2bag/ros2bag/verb/play.py:27]: This is an API/workflow breaking change. Why is this necessary?

File: ros2bag/ros2bag/verb/play.py [hunk 2]
```diff
@@ -47,7 +52,7 @@ def check_positive_float(self, value):
             raise ArgumentTypeError('%s is not of the valid type (float)' % value)
 
     def main(self, *, args):  # noqa: D102
-        bag_file = args.bag_file
+        bag_file = args.input
```
Comment (piraka9011) [inline ros2bag/ros2bag/verb/play.py:55]: ditto

File: rosbag2_transport/include/rosbag2_transport/play_options.hpp [hunk 3]
```diff
@@ -27,6 +28,7 @@ struct PlayOptions
   size_t read_ahead_queue_size;
   std::string node_prefix = "";
   float rate = 1.0;
+  std::vector<std::string> topics = std::vector<std::string>();
```
Comment (piraka9011) [inline rosbag2_transport/include/rosbag2_transport/play_options.hpp:31]: `topics_to_filter` or a more descriptive name.

File: rosbag2_transport/src/rosbag2_transport/rosbag2_transport_python.cpp [hunk 4]
```diff
@@ -181,6 +184,19 @@ rosbag2_transport_play(PyObject * Py_UNUSED(self), PyObject * args, PyObject * k
   play_options.read_ahead_queue_size = read_ahead_queue_size;
   play_options.rate = rate;
 
+  if (topics) {
+    PyObject * topic_iterator = PyObject_GetIter(topics);
+    if (topic_iterator != nullptr) {
+      PyObject * topic;
```
Comment (piraka9011) [inline rosbag2_transport/src/rosbag2_transport/rosbag2_transport_python.cpp:190]:
```suggestion
      PyObject * topic = nullptr;
```

File: rosbag2_transport/src/rosbag2_transport/rosbag2_transport_python.cpp [hunk 5]
```diff
@@ -181,6 +184,19 @@ rosbag2_transport_play(PyObject * Py_UNUSED(self), PyObject * args, PyObject * k
...
+      while ((topic = PyIter_Next(topic_iterator))) {
```
Comment (piraka9011) [inline rosbag2_transport/src/rosbag2_transport/rosbag2_transport_python.cpp:191]: Can this be simplified?

File: ros2bag/ros2bag/verb/play.py (hunk 1 above)
Comment (piraka9011) [inline ros2bag/ros2bag/verb/play.py:27]: Is this necessary?

Comment (mabelzhang) [reply ros2bag/ros2bag/verb/play.py]: Artifact from previous testing. Removed in 7e7eb9b

File: rosbag2_transport/include/rosbag2_transport/play_options.hpp (hunk 3 above)
Comment (mabelzhang) [reply rosbag2_transport/include/rosbag2_transport/play_options.hpp]: 7e7eb9b

File: ros2bag/ros2bag/verb/play.py [hunk 6]
```diff
@@ -47,7 +51,7 @@ def check_positive_float(self, value):
...
```
Comment (piraka9011) [inline ros2bag/ros2bag/verb/play.py:54]: Don't forget this too

File: rosbag2_transport/src/rosbag2_transport/rosbag2_transport_python.cpp (hunk 5 above)
Comment (mabelzhang) [reply rosbag2_transport/src/rosbag2_transport/rosbag2_transport_python.cpp:236]: I don't know enough about Python C to simplify it with confidence that I'm not introducing errors. I copied it from the record command's processing of topics [above](https://github.com/ros2/rosbag2/blob/master/rosbag2_transport/src/rosbag2_transport/rosbag2_transport_python.cpp#L153). It seems that vectors of strings have to be treated differently when converting from Python. I don't know why it has to do `PYDECREF` as opposed to the normal way to loop...

File: ros2bag/ros2bag/verb/play.py (hunk 6 above)
Comment (mabelzhang) [reply ros2bag/ros2bag/verb/play.py]: Haha good catch. I luckily deleted it when I merged from master.

File: ros2bag/ros2bag/verb/play.py [hunk 7]
```diff
@@ -42,6 +42,9 @@ def add_arguments(self, parser, cli_name):  # noqa: D102
             '-r', '--rate', type=check_positive_float, default=1.0,
             help='rate at which to play back messages. Valid range > 0.0.')
         parser.add_argument(
+            '--topics', type=str, default='', nargs='*',
```
Comment (Karsten1987) [inline ros2bag/ros2bag/verb/play.py:45]: shouldn't that be `nargs='+'`? Otherwise I could call `ros2 bag play --topics` and don't specify any topics. I think that's somewhat weird.

File: rosbag2_tests/test/rosbag2_tests/test_rosbag2_play_end_to_end.cpp [hunk 8]
```diff
@@ -118,3 +118,90 @@ TEST_F(PlayEndToEndTestFixture, play_fails_gracefully_if_needed_coverter_plugin_
...
+  subscription_future = sub_->spin_subscriptions();
+
+  exit_code = execute_and_wait_until_completion("ros2 bag play --topics /array_topic -- cdr_test",
+      database_path_);
+
+  subscription_future.get();
+
+  primitive_messages = sub_->get_received_messages<test_msgs::msg::BasicTypes>("/test_topic");
+  array_messages = sub_->get_received_messages<test_msgs::msg::Arrays>(
+    "/array_topic");
```
Comment (Karsten1987) [inline rosbag2_tests/test/rosbag2_tests/test_rosbag2_play_end_to_end.cpp:158]: nit: why having the line break here?

File: rosbag2_transport/include/rosbag2_transport/play_options.hpp [hunk 9]
```diff
@@ -30,6 +31,12 @@ struct PlayOptions
   size_t read_ahead_queue_size;
   std::string node_prefix = "";
   float rate = 1.0;
+
+  // Topic names to whitelist when playing a bag. Only messages matching these
+  // specified topics will be played. If list is empty, the filter is ignored
+  // and all messages are played.
```
Comment (Karsten1987) [inline rosbag2_transport/include/rosbag2_transport/play_options.hpp:37]: we generally have the rule of having one sentence per line

File: rosbag2_transport/include/rosbag2_transport/play_options.hpp [hunk 10]
```diff
@@ -30,6 +31,12 @@ struct PlayOptions
...
+  std::vector<std::string> topics_to_filter = std::vector<std::string>();
```
Comment (Karsten1987) [inline rosbag2_transport/include/rosbag2_transport/play_options.hpp:38]:
```suggestion
  std::vector<std::string> topics_to_filter ={};
```

Comment (emersonknapp) [reply rosbag2_transport/include/rosbag2_transport/play_options.hpp:38]: `std::vector<std::string> topics_to_filter{}`

Comment (piraka9011) [reply rosbag2_transport/include/rosbag2_transport/play_options.hpp:38]: Or just `std::vector<std::string> topics_to_filter{};

Comment (Karsten1987) [reply rosbag2_transport/include/rosbag2_transport/play_options.hpp]: just adapting to the style below for the `std::unordered_map` ;-)

File: ros2bag/ros2bag/verb/play.py (hunk 7 above)
Comment (mabelzhang) [reply ros2bag/ros2bag/verb/play.py]: 520fd60

File: rosbag2_tests/test/rosbag2_tests/test_rosbag2_play_end_to_end.cpp (hunk 8 above)
Comment (mabelzhang) [reply rosbag2_tests/test/rosbag2_tests/test_rosbag2_play_end_to_end.cpp:158]: I just copied the format from the play_end_to_end test above for these configurations.

File: rosbag2_transport/include/rosbag2_transport/play_options.hpp (hunk 9 above)
Comment (mabelzhang) [reply rosbag2_transport/include/rosbag2_transport/play_options.hpp]: 520fd60

File: rosbag2_tests/test/rosbag2_tests/test_rosbag2_play_end_to_end.cpp (hunk 8 above)
Comment (emersonknapp) [reply rosbag2_tests/test/rosbag2_tests/test_rosbag2_play_end_to_end.cpp:158]: Probably copy-paste is not a good reason for code to be any particular way.

**Comment by mabelzhang** [comment]
My local workspace is very broken. I need to fix it before I can compile locally before trying CI again.

File: rosbag2_tests/test/rosbag2_tests/test_rosbag2_play_end_to_end.cpp (hunk 8 above)
Comment (mabelzhang) [reply rosbag2_tests/test/rosbag2_tests/test_rosbag2_play_end_to_end.cpp]: Oh I misread. I thought the comment is talking about the empty lines. The line break is not necessary. Removed line break in 5bd47bc

**Comment by emersonknapp** [comment]
I recommend a docker-based workflow. Here's the Dockerfile I personally use https://github.com/emersonknapp/ros2containerbuild/blob/master/Dockerfile and use via

```
rocker \
  --x11 \
  --network host \
  --oyr-run-arg "--name SOMETHING --rm -v $PWD:/ws -w /ws" \
  candleends/ros2dev
```

Workspace always clean and portable :)

File: rosbag2_transport/include/rosbag2_transport/play_options.hpp (hunk 10 above)
Comment (mabelzhang) [reply rosbag2_transport/include/rosbag2_transport/play_options.hpp]: Heh, so I already changed to `{}` without the `=` so now they're inconsistent. Do we prefer changing the line below too, or change back to `={}`?

**Comment by mabelzhang** [comment]
Yeah... I do develop in Docker containers normally, but the one I've been using is built from eloquent source, and with the recent volume of changes, I'm not able to just pull a few packages from master branch and compile. So I'm making a new Docker image, probably not unlike yours, that pulls ros2.repos from the master branch instead.

**Comment by emersonknapp** [comment]
I did not intend to close this... no idea how i did

**Comment by mabelzhang** [comment]
Darn. I thought there was no more work for me ;)

**Comment by mabelzhang** [comment]
There are two CI test failures on Linux that I don't understand, `projectroot.test_record_all__rmw_fastrtps_cpp` and `rosbag2_transport.RecordIntegrationTestFixture.published_messages_from_multiple_topics_are_recorded`. They're printing binary bites. Still waiting for macOS and Windows.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10191/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5744/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8283/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10076/)

**Comment by mabelzhang** [comment]
macOS has the same test failures as Linux, and Windows has some record test failures.

**Comment by emersonknapp** [comment]
See https://github.com/ros2/rosbag2/pull/376#issuecomment-613711587 - my current hypothesis is that there is a bug in Fast-RTPS right now...

**Comment by mabelzhang** [comment]
No more test failures in Linux and macOS. Waiting for Windows
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10229/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5777/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8318/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10117/)

**Comment by mabelzhang** [comment]
Windows tests look okay too. Only the known record split bag failures.

File: rosbag2_transport/include/rosbag2_transport/play_options.hpp [hunk 11]
```diff
@@ -17,6 +17,7 @@
 
 #include <cstddef>
 #include <string>
+#include <vector>
```
Comment (Karsten1987) [inline rosbag2_transport/include/rosbag2_transport/play_options.hpp:20]: alpha

Comment (mabelzhang) [reply rosbag2_transport/include/rosbag2_transport/play_options.hpp]: ff5de2f
